File: LLM-Translations/GPT-4/avatar/Python/C/atcoder_AGC046_B.c
#include <stdio.h>

int main() {
    int A, B, C, D;
    scanf("%d %d %d %d", &A, &B, &C, &D);
    int m = 998244353;
    int d[C + 1][D + 1];
    for(int i = 0; i <= C; i++) {
        for(int j = 0; j <= D; j++) {
            d[i][j] = 0;
        }
    }
    d[A][B] = 1;
    for(int i = A; i <= C; i++) {
        for(int j = B; j <= D; j++) {
            d[i][j] = (d[i][j] + ((d[i][j - 1] * i + d[i - 1][j] * j - d[i - 1][j - 1] * (i - 1) * (j - 1)) % m)) % m;
        }
    }
    printf("%d", d[C][D]);
    return 0;
}
//End of Code

File: LLM-Translations/GPT-4/avatar/Python/Java/atcoder_AGC046_B.java

import java.util.Scanner;

public class atcoder_AGC046_B {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int A = in.nextInt();
        int B = in.nextInt();
        int C = in.nextInt();
        int D = in.nextInt();
        in.close();

        int m = 998244353;
        int[][] d = new int[C + 1][D + 1];
        d[A][B] = 1;

        for (int i = A; i <= C; i++) {
            for (int j = B; j <= D; j++) {
                d[i][j] += (int)(((long)d[i][j - 1] * i + (long)d[i - 1][j] * j - (long)d[i - 1][j - 1] * (i - 1) * (j - 1)) % m);
            }
        }

        System.out.println(d[C][D]);
    }
}

// End of Code.
